storage_controller/src/service.rs

```diff
@@ -4952,4 +5139,253 @@ impl Service {
+
+                let can_take = attached - expected_attached;
+                let mut remove_node = false;
+                for _ in 0..can_take {
+                    match tids_by_node.get_mut(&node_id) {
+                        Some(tids) => match tids.pop() {
+                            Some(tid) => {
+                                plan.push(tid);
+                            }
+                            None => {
+                                remove_node = true;
+                                break;
+                            }
+                        },
+                        None => {
+                            break;
+                        }
+                    }
+                }
+
+                if remove_node {
+                    tids_by_node.remove(&node_id);
+                }
+            }
+
+            plan
+        };
+
+        let mut waiters = Vec::new();
+        let mut schedule_context = ScheduleContext::default();
+
+        // Execute the plan we've composed above. Before aplying each move from the plan,
+        // we validate to ensure that it has not gone stale in the meantime.
+        while !tids_to_promote.is_empty() {
+            if cancel.is_cancelled() {
+                return Err(OperationError::Cancelled);
+            }
+
+            {
+                let mut locked = self.inner.write().unwrap();
+                let (nodes, tenants, scheduler) = locked.parts_mut();
+
+                let node = nodes.get(&node_id).ok_or(OperationError::NodeStateChanged(
+                    format!("node {node_id} was removed").into(),
+                ))?;
+
+                let current_policy = node.get_scheduling();
+                if !matches!(current_policy, NodeSchedulingPolicy::Filling) {
+                    // TODO(vlad): maybe cancel pending reconciles before erroring out. need to think
+                    // about it
+                    return Err(OperationError::NodeStateChanged(
+                        format!("node {node_id} changed state to {current_policy:?}").into(),
+                    ));
+                }
+
+                while waiters.len() < MAX_RECONCILES_PER_OPERATION {
+                    if let Some(tid) = tids_to_promote.pop() {
+                        if let Some(tenant_shard) = tenants.get_mut(&tid) {
+                            // If the node being filled is not a secondary anymore,
+                            // skip the promotion.
+                            if !tenant_shard.intent.get_secondary().contains(&node_id) {
+                                continue;
+                            }
+
+                            tenant_shard.intent.promote_attached(scheduler, node_id);
+                            match tenant_shard.schedule(scheduler, &mut schedule_context) {
+                                Err(e) => {
+                                    tracing::warn!(%tid, "Scheduling error when filling pageserver {} : {e}", node_id);
```

Comment:
We already log the tenant shard id (`%tid`). Is this what you meant?

Reply:
Argh, my eyes just skipped over the `%tid` because it's little.

Most of the logs use `tenant_id=%tenant_shard_id.tenant_id, shard_id=%tenant_shard_id.shard_slug()` -- which is a mouthful, but enables folks to filter logs on an exact tenant ID field.

Reply:
Make sense. Done in [627cd0c](https://github.com/neondatabase/neon/pull/8014/commits/627cd0c0992756b6459367b725c87cd3e35e451b)